README.md
### GobCrypt

#### A Bcrypt hash password generator written in golang

[![test-lint](https://github.com/mijho/gobcrypt/actions/workflows/test-lint.yml/badge.svg?branch=main)](https://github.com/mijho/gobcrypt/actions/workflows/test-lint.yml)
[![build-release](https://github.com/mijho/gobcrypt/actions/workflows/build-release.yml/badge.svg?branch=main)](https://github.com/mijho/gobcrypt/actions/workflows/build-release.yml)
![go-report](https://goreportcard.com/badge/github.com/mijho/gobcrypt)

##### Installation Instructions

With go get:
```
go get github.com/mijho/gobcrypt
```

From source:
```
git clone github.com/mijho/gobcrypt.git
make build
```

Or head to the releases page to download the prebuild binary for your system.

## Pass/Hash Generation

```
NAME:
   gobcrypt generate - generate pass/hash pairs

USAGE:
   gobcrypt generate [command options] [arguments...]

OPTIONS:
   --number value, -n value       Specify the number of hashes to create (default: 1) (default: 1)
   --password value, -p value     Password to hash
   --length value, -l value       Specify the length of password required (default: 16) (default: 16)
   --input-file value, -i value   Specify a file to read passwords from
   --output-file value, -o value  Specify a file to write out the pass/hash to
   --cost value, -c value         Specify the cost to use (Min: 4, Max: 31) (default: 14) (default: 0)
   --help, -h                     show help (default: false)
```

## Pass/Hash Validation

```
NAME:
   gobcrypt validate - validate pass/hash pairs

USAGE:
   gobcrypt validate [command options] [arguments...]

OPTIONS:
   --password value, -p value     Password to hash
   --hash value, --hp value       Hash to verify
   --input-file value, -i value   Specify a file to read passwords from
   --output-file value, -o value  Specify a file to write out the pass/hash to
   --help, -h                     show help (default: false)
```

## Examples

Create a password and hash:
```
$ ./gobcrypt generate
f9fFFh7LcAaGJ89C $2a$10$pPbKZkrpHmJg5vZVGljUV.3N/gatYJN4Iv3bAnJ7b7SKgQTfZCnhS
```
Create a hash of a specific password
```
$ ./gobcrypt generate -p supergoodpass
supergoodpass $2a$10$KGm2jtacLvCXT2uitl8aQOPiDZBX8VmXIS9dZS5aAtQsQAVYetwiG
```

Check the hash is valid:
```
$ ./gobcrypt validate -p supergoodpass -hp '$2a$10$KGm2jtacLvCXT2uitl8aQOPiDZBX8VmXIS9dZS5aAtQsQAVYetwiG'
MATCH: PASS, password: supergoodpass, hash: $2a$10$KGm2jtacLvCXT2uitl8aQOPiDZBX8VmXIS9dZS5aAtQsQAVYetwiG
```

Create passwords of a specific length (15 chars by default)
```
$ ./gobcrypt generate -l 20                                                                              
6DG4D7WZnA86kD2iQnnD $2a$10$IHrYPjzwBDG2BYTIcOBFlutNHEB2mBmpP6QG8ZF4iCW8Bn3suToMu
```

Create a bulk amount of password hash combinations
```
$ ./gobcrypt generate -n 7
EQKEBdAYXPZNwMM0 $2a$10$.iB7vbaKj2L7cYpTTa/.nuHe/BKQEL3uXgqduBOpGVhvuHkSz0yoS
y8mEjf6uIYx3QyLs $2a$10$QpSqInArVO9A5s/vpkgW1eOgtdLV2TJw1dqD3aQjYF0kbX0FpDp/e
akA95A1qiwmoPcjw $2a$10$RhKxTNn.2CPB7ertSScKFO30Fqcg9xAJH.J2X8PspWLm31D2O1wR6
J5H58PoVxR6j8n3s $2a$10$WAENUArAzRSo8zXgFbB4COUKv6vhDbPyTkFIqXVr5JOf5ksjspC2K
o1xb8hiqPrXhEmCT $2a$10$Io7hlQuc2C1Z9A4IxVDSxO6cx2SCql1Gb65/YCb4Xj.djeGa1bDSS
mkI6LG4Yc0d4drvE $2a$10$L.76ZOQNZkhbjQX7GzupzO4Ke/VEdlCwwv4q5vzHX5/ojEH95RkPi
t3LjJcaHg2nEY6mZ $2a$10$mduGQTccWkQ9dQ6n1LlBxeII1IlfD6uB51Ewy/MTVQ0q88DtMHTjO
```

Create a bulk amount and write the output to file
```
$ ./gobcrypt generate -o outfile.txt && cat outfile.txt                                                 
uLTAPhWdIqGq3vPY $2a$10$zCb9z3melGqqw8I1aancVeMBS7xzXIaAAGsaDbqpj/AO3VP1/XNEi
$ ./gobcrypt validate -p supergoodpass -hp '$2a$10$KGm2jtacLvCXT2uitl8aQOPiDZBX8VmXIS9dZS5aAtQsQAVYetwiG' -o results.txt && cat results.txt
MATCH: PASS, password: supergoodpass, hash: $2a$10$KGm2jtacLvCXT2uitl8aQOPiDZBX8VmXIS9dZS5aAtQsQAVYetwiG
```

Read passwords from file to hash or validate
```
$ cat infile.txt
password
hotdamn
12hfsklfh
forbarbaz
$ ./gobcrypt generate -i infile.txt                        
password $2a$10$7Ttcr9C/6QkBnyewt/Ljg.HRpfoil96XSjTve7EXy5XOBiZB5Zr4a
hotdamn $2a$10$072dU17Zf/M.SbjAP.zhNer39Z7fPOUWHOTm6.849OzA9f3NJFaGu
12hfsklfh $2a$10$tVyXGyJ674PVJlh.dDd.seOA/RJRbDzPAlxFNjKp1Z3mVpOtOTJIG
forbarbaz $2a$10$bboUUJP0Fdp9oWBSFGN.GuV/VfBgYuHQfpQ3JcElzMH7ZeVCes8w2
$ ./gobcrypt validate -i outfile.txt 
MATCH: PASS, password: uLTAPhWdIqGq3vPY, hash: $2a$10$zCb9z3melGqqw8I1aancVeMBS7xzXIaAAGsaDbqpj/AO3VP1/XNEi
```

Manually set the cost of the BCrypt hashing process
```
$ ./gobcrypt generate -c 17
NKOSv14NuLViUqkf $2a$17$DWJWvyIZH/CamvGAa59Naefhz5AnFNFbfkQ38ryhsH27J7Zovt8Gm
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [unreleased]

### Bug Fixes

- Fmt -> log statements

### Documentation

- Updated readme and add changelog

### Features

- Make bcrypt cost configurable
- Add validate cmd option

### Miscellaneous Tasks

- Add go mod
- Remove old ci and go files
- Update changelog
- Merge conflict
- Bump version
- Add comment to CLIApp

### Refactor

- Add in urfave cli and first handler
- Basic funcitonality ported
- Move output to util func

### Ci

- Add in changelog configuration
- Add in debug for ci env
- Update changelog
- Remove debug env
- Split out changelog ci
- Update changelog
- Set up CI for release generation
- Update changelog
- Rename build-release flow
- Update changelog
- Remove default test
- Update changelog
- Update changelog
- Update release to use custom pat
- Update changelog
- Update changelog
- Set release tag as env
- Update changelog
- Set release tag as env
- Update changelog
- Update go report
- Add some badges
- Update changelog
- Update changelog
- Move go report to build-release

<!-- generated by git-cliff -->


cmd/utils_test.go
package main

import (
	"testing"

	"golang.org/x/crypto/bcrypt"
)

func Test_generateHashForPassword(t *testing.T) {
	password := "superSecretPassword123"
	hashLine, _ := generateHashForPassword(password, 14)
	if err := checkHashAndPassword(hashLine[0]); err != nil {
		t.Error("Hash and Password do not match")
	}
}

func Test_generateHashForRandomPassword(t *testing.T) {
	password := randomString(20)
	hashLine, _ := generateHashForPassword(password, 14)
	if err := checkHashAndPassword(hashLine[0]); err != nil {
		t.Error("Hash and Password do not match")
	}
}

func Test_InvalidHashAndPassword(t *testing.T) {
	password := "password"
	hash := "$2a$14$9/5dnFqcgsFRWYEOWi3iyiyiaPOBX6yEJA4vNZIK78cUIOuryK3vMPYfSWs"
	err := bcrypt.CompareHashAndPassword([]byte(hash), []byte(password))
	if err == nil {
		t.Errorf("Password and Hash should not match: %v", err)
	}
}

func Test_RandomPasswordLength(t *testing.T) {
	password := randomString(20)
	if len(password) != 20 {
		t.Error("Generated Password not correct length")
	}
}


gobcrypt.go
// gobcrypt.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"golang.org/x/crypto/bcrypt"
	"log"
	"math/rand"
	"os"
	"strings"
	"time"
	"strconv"
)

var r *rand.Rand // Rand for this package.

func init() {
	r = rand.New(rand.NewSource(time.Now().UnixNano()))
}

/* This function will generate a random string based on the random value
*  set from the init function ant the chars defined below
 */
func randomString(strlen int) string {
	const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
	result := ""
	for i := 0; i < strlen; i++ {
		index := r.Intn(len(chars))
		result += chars[index : index+1]
	}
	return result
}

/* This function will create a hashed password using bcrypt. It requires a
*  String value to be passed as an argument and returns the hash as a String
 */
func hashPassword(password string) (string, error) {
	bytes, err := bcrypt.GenerateFromPassword([]byte(password), 14)
	return string(bytes), err
}

/* This function compares the plain text password with the hash to ensure it is
*  valid. It requires both the plain text and hash pass String to be passed to it
 */
func checkHashAndPassword(password, hash string) bool {
	err := bcrypt.CompareHashAndPassword([]byte(hash), []byte(password))
	return err == nil
}

// readLines reads a whole file into memory
// and returns a slice of its lines.
func readLines(path string) ([]string, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

// writeLines writes the lines to the given file.
func writeLines(lines []string, path string) error {
	file, err := os.Create(path)
	if err != nil {
		return err
	}
	defer file.Close()

	w := bufio.NewWriter(file)
	for _, line := range lines {
		fmt.Fprintln(w, line)
	}
	return w.Flush()
}

/* Processes the returned Bool value from checkHashAndPassword
*  and formats it for use in hashLines
*/
func matchPasswordAndHash(password, hash string) string {
	match := checkHashAndPassword(password, hash)
	matchString := strconv.FormatBool(match)
	matchArray := []string{"Match: ", matchString}
	matchLine := strings.Join(matchArray, " ")
	return string(matchLine)
}

func main() {

	var count, pwLength int
	var thisPW, inFile, outFile string
	var testHash bool
	flag.IntVar(&count, "c", 1, "Specify the number of hashes to create")
	flag.StringVar(&thisPW, "s", "", "Hash the specified password only (1)")
	flag.BoolVar(&testHash, "t", false, "Validate the hash & pass")
	flag.IntVar(&pwLength, "l", 15, "Specify the length of password required")
	flag.StringVar(&inFile, "f", "", "Specify a file to read passwords from")
	flag.StringVar(&outFile, "o", "", "Specify a file to write out the pass/hash to")
	flag.Parse()

	if thisPW == "" {

		if inFile != "" {

			lines, err := readLines(inFile)
			
			if err != nil {
				log.Fatalf("readLines: %s", err)
			}
			var hashLines []string
			
			for _, line := range lines {
				
				password := line
				hash, _ := hashPassword(password)
				var matchLine string
				hashArray := []string{password, hash}
				hashLine := strings.Join(hashArray, "	")
				hashLines = append(hashLines, hashLine)

				if testHash != false {
					matchLine = matchPasswordAndHash(password, hash)
					hashLines = append(hashLines, matchLine)
				}
				
				if outFile != "" {
					if err := writeLines(hashLines, outFile); err != nil {
						log.Fatal("writeLine: %s", err)
					}
				} else {
					fmt.Println(hashLine)
					if testHash != false {
						fmt.Println(matchLine)
					}
				}
			}
		} else {

			var matchLine string
			var hashLines []string

			for count > 0 {

				password := randomString(pwLength)
				hash, _ := hashPassword(password) // TODO add error handling
				hashArray := []string{password, hash}
				hashLine := strings.Join(hashArray, "	")
				hashLines = append(hashLines, hashLine)

				if testHash != false {
					matchLine = matchPasswordAndHash(password, hash)
					hashLines = append(hashLines, matchLine)
				}
				if outFile == "" {
					fmt.Println(hashLine)
					if testHash != false {
						fmt.Println(matchLine)
					}
				}
				count -= 1
			}
			if outFile != "" {
				if err := writeLines(hashLines, outFile); err != nil {
					log.Fatal("writeLine: %s", err)
				}
			} 			
		}
	} else {

		if inFile != "" {
			log.Fatalf("Cannot run -s and -f together please select one")
		}

		var matchLine string
		var hashLines []string
		password := thisPW
		hash, _ := hashPassword(password)
		hashArray := []string{password, hash}
		hashLine := strings.Join(hashArray, "	")
		hashLines = append(hashLines, hashLine)

		if testHash != false {
			matchLine = matchPasswordAndHash(password, hash)
			hashLines = append(hashLines, matchLine)
		}
		if outFile != "" {
			if err := writeLines(hashLines, outFile); err != nil {
				log.Fatal("writeLine: %s", err)
			}
		} else {
			fmt.Println(hashLine)
			if testHash != false {
				fmt.Println(matchLine)
			}
		}
	}
}


cmd/utils.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strings"
	"time"

	"golang.org/x/crypto/bcrypt"
)

var r *rand.Rand // Rand for this package.

func init() {
	r = rand.New(rand.NewSource(time.Now().UnixNano()))
}

// This function will generate a random string based on the random value
//  set from the init function ant the chars defined below
func randomString(strlen int) string {
	const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
	result := ""
	for i := 0; i < strlen; i++ {
		index := r.Intn(len(chars))
		result += chars[index : index+1]
	}
	return result
}

/* This function will create a hashed password using bcrypt. It requires a
*  String value to be passed as an argument and returns the hash as a String
 */
func hashPassword(password string, cost int) (string, error) {
	bytes, err := bcrypt.GenerateFromPassword([]byte(password), cost)
	return string(bytes), err
}

/* This function compares the plain text password with the hash to ensure it is
*  valid. It requires both the plain text and hash pass String to be passed to it
 */
func checkHashAndPassword(hashLine string) error {
	items := strings.Split(hashLine, " ")
	hash, password := items[1], items[0]
	err := bcrypt.CompareHashAndPassword([]byte(hash), []byte(password))
	return err
}

// readLines reads a whole file into memory
// and returns a slice of its lines.
func readLines(path string) ([]string, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

// writeLines writes the lines to the given file.
func writeLines(lines []string, path string) error {
	file, err := os.Create(path)
	if err != nil {
		return err
	}
	defer file.Close()

	w := bufio.NewWriter(file)
	for _, line := range lines {
		fmt.Fprintln(w, line)
	}
	return w.Flush()
}

// hanldes returning the output to user via file or stdout
func returnOutput(hashLines []string, outFile string) error {
	if outFile != "" {
		if err := writeLines(hashLines, outFile); err != nil {
			return err
		}
	} else {
		for _, line := range hashLines {
			fmt.Println(line)
		}
	}
	return nil
}

// generates a hash from the given password string
func generateHashForPassword(password string, cost int) ([]string, error) {
	var hashLines []string

	hash, err := hashPassword(password, cost)
	if err != nil {
		return nil, err
	}

	hashLine := strings.Join([]string{password, hash}, " ")
	hashLines = append(hashLines, hashLine)

	return hashLines, err
}

// Processes the returned Bool value from checkHashAndPassword
//  and formats it for use in hashLines
func matchPasswordAndHash(hashLine string) (result string, err error) {
	items := strings.Split(hashLine, " ")
	hash, password := items[1], items[0]
	err = checkHashAndPassword(hashLine)
	if err != nil {
		result = fmt.Sprintf("MATCH: FAIL, password: %s, hash: %s", password, hash)
		return result, err
	}
	result = fmt.Sprintf("MATCH: PASS, password: %s, hash: %s", password, hash)
	return result, err
}


cmd/cli.go
package main

import (
	"fmt"
	"log"
	"strings"

	"github.com/urfave/cli/v2"
)

// Build time variables
var (
	Build   string
	Commit  string
	Name    string
	Version string
	Author  string
	Email   string
)

var numberValue int
var numberFlag = &cli.IntFlag{
	Name:        "number",
	Aliases:     []string{"n"},
	Usage:       "Specify the number of hashes to create (default: 1)",
	Value:       1,
	Required:    false,
	Destination: &numberValue,
}

var passwordValue string
var passwordFlag = &cli.StringFlag{
	Name:        "password",
	Aliases:     []string{"p"},
	Usage:       "Password to hash",
	Required:    false,
	Destination: &passwordValue,
}

var hashValue string
var hashFlag = &cli.StringFlag{
	Name:        "hash",
	Aliases:     []string{"hp"},
	Usage:       "Hash to verify",
	Required:    false,
	Destination: &hashValue,
}

var lengthValue int
var lengthFlag = &cli.IntFlag{
	Name:        "length",
	Aliases:     []string{"l"},
	Usage:       "Specify the length of password required (default: 16)",
	Value:       16,
	Required:    false,
	Destination: &lengthValue,
}

var inputFileValue string
var inputFileFlag = &cli.StringFlag{
	Name:        "input-file",
	Aliases:     []string{"i"},
	Usage:       "Specify a file to read passwords from",
	Required:    false,
	Destination: &inputFileValue,
}

var outputFileValue string
var outputFileFlag = &cli.StringFlag{
	Name:        "output-file",
	Aliases:     []string{"o"},
	Usage:       "Specify a file to write out the pass/hash to",
	Required:    false,
	Destination: &outputFileValue,
}

var costValue int
var costFlag = &cli.IntFlag{
	Name:        "cost",
	Aliases:     []string{"c"},
	Usage:       "Specify the cost to use (Min: 4, Max: 31) (default: 14)",
	Required:    false,
	Destination: &costValue,
}

// CLIApp constructs the cli applications
var CLIApp = &cli.App{
	Name:        "gobcrypt",
	Usage:       "A Bcrypt hash/password generator",
	ArgsUsage:   "",
	Version:     Version + "-" + Commit,
	Description: fmt.Sprintf("%s: Build: %s", Name, Build),
	Commands: []*cli.Command{
		{
			Name:  "generate",
			Usage: "generate pass/hash pairs",
			Flags: []cli.Flag{
				numberFlag,
				passwordFlag,
				lengthFlag,
				inputFileFlag,
				outputFileFlag,
				costFlag,
			},
			Action: GenerateHandler,
		},
		{
			Name:  "validate",
			Usage: "validate pass/hash pairs",
			Flags: []cli.Flag{
				passwordFlag,
				hashFlag,
				inputFileFlag,
				outputFileFlag,
			},
			Action: ValidateHandler,
		},
	},
	Authors: []*cli.Author{
		{
			Name:  Author,
			Email: Email,
		},
	},
}

// ValidateHandler provides functionality to validate pass/hash pairs
func ValidateHandler(c *cli.Context) error {
	var hashLines []string

	if inputFileValue != "" {
		lines, err := readLines(inputFileValue)

		if err != nil {
			log.Fatalf("There was an error reading lines from file: %s\n", err)
		}
		for _, line := range lines {
			result, _ := matchPasswordAndHash(line)
			hashLines = append(hashLines, result)
		}
		if err := returnOutput(hashLines, outputFileValue); err != nil {
			log.Fatalf("There was an error writing lines to file: %s\n", err)
		}
	} else {
		hashLine := strings.Join([]string{passwordValue, hashValue}, " ")
		result, _ := matchPasswordAndHash(hashLine)
		hashLines = append(hashLines, result)

		if err := returnOutput(hashLines, outputFileValue); err != nil {
			log.Fatalf("There was an error writing lines to file: %s\n", err)
		}
	}

	return nil
}

// GenerateHandler provides functionality to generate pass/hash pairs
func GenerateHandler(c *cli.Context) error {
	cost := costValue

	if passwordValue != "" {
		hashLines, err := generateHashForPassword(passwordValue, cost)
		if err != nil {
			log.Fatalf("There was an error creating a hash: %s\n", err)
		}

		if err := returnOutput(hashLines, outputFileValue); err != nil {
			log.Fatalf("There was an error writing lines to file: %s\n", err)
		}
		return nil
	}

	if inputFileValue != "" {
		lines, err := readLines(inputFileValue)

		if err != nil {
			log.Fatalf("There was an error reading lines from file: %s\n", err)
		}

		var hashLines []string
		for _, password := range lines {
			hashLine, err := generateHashForPassword(password, cost)
			if err != nil {
				log.Fatalf("There was an error creating a hash: %s\n", err)
			}
			hashLines = append(hashLines, hashLine...)
		}

		if err := returnOutput(hashLines, outputFileValue); err != nil {
			log.Fatalf("There was an error writing lines to file: %s\n", err)
		}
		return nil
	}

	var hashLines []string
	pwLength := lengthValue
	for number := numberValue; number > 0; number-- {
		password := randomString(pwLength)
		hashLine, err := generateHashForPassword(password, cost)
		if err != nil {
			log.Fatalf("There was an error creating a hash: %s\n", err)
		}
		hashLines = append(hashLines, hashLine...)
	}

	if err := returnOutput(hashLines, outputFileValue); err != nil {
		log.Fatalf("There was an error writing lines to file: %s\n", err)
	}

	return nil
}
